### ISSUE TITLE
vllm backend gives different results from LanguageModel

### ISSUE THREAD

**Issue by wendlerc** [ndif-team/nnsight#337]
![Image](https://github.com/user-attachments/assets/d9d5c872-8a5a-4920-9913-b76f1a4dcb3c)

Snipped for reproducing the behavior attached.

```
# %%
from nnsight.modeling.vllm import VLLM
from nnsight import LanguageModel 
import torch

llm = LanguageModel("deepseek-ai/DeepSeek-R1-Distill-Llama-8B",
                    device_map="auto", 
                    torch_dtype=torch.bfloat16)



# %%
with llm.trace("this is a test") as trace:
        latents = llm.model.layers[14].output[0].save()
del llm

# %%
vllm = VLLM("deepseek-ai/DeepSeek-R1-Distill-Llama-8B",
            device="auto", 
            dispatch=True, 
            dtype="bfloat16",
            max_model_len=15000)

# %%
with vllm.trace("this is a test") as trace:
        vlatents = vllm.model.layers[14].output[0].save()
del vllm

# %%
print((latents - vlatents).abs().mean())
```

**Comment by wendlerc** [comment]
Here is another snippet that might be helpful to get at the bottom of this. When grabbing latents after the layer_norm layers both llm and vllm actually agree. However, I did not manage to reproduce the normalized latents in the case of vllm by applying the steps of their rmsnorm implementation manually.

```
from nnsight.modeling.vllm import VLLM
from nnsight import LanguageModel 
import torch

llm = LanguageModel("/tmp/DeepSeek-R1-Distill-Llama-8B",
                    device_map="auto", 
                    torch_dtype=torch.bfloat16)

# this needs to be there such that vllm does not use up all the memory.
with llm.trace("this is a test") as trace:
    residual = llm.model.layers[0].output[0].save()
    norm = llm.model.layers[1].input_layernorm.output[0].save()

vllm = VLLM("/tmp/DeepSeek-R1-Distill-Llama-8B",
            device="auto", 
            dispatch=True, 
            dtype="bfloat16",
            max_model_len=1000)

# input layer norm
with llm.trace("this is a test") as trace:
    residual = llm.model.layers[1].output[0].save()
    norm = llm.model.layers[2].input_layernorm.output[0].save()
with vllm.trace("this is a test") as trace:
    vresidual = vllm.model.layers[1].output.save()
    vnorm = vllm.model.layers[2].input_layernorm.output[0].save()
print("the output of the previous layer / input to the current layer are different")
print((residual - vresidual[0]).abs().mean())
print((residual - vresidual[1]).abs().mean())
#print(residual)
#print(vresidual[0])
#print(vresidual[1])
print("the normalized inputs are the same")
print((norm - vnorm).abs().mean())
#print(norm)
#print(vnorm)

# trying to compute the layer norm manually
# trying to apply layernorm manually fails to produce the same result in the case of vllm
# apply layernorm manually (before they are different, after they are same...)
mynorm = residual.float() * torch.rsqrt(residual.float().pow(2).mean(dim=-1, keepdim=True) + 1e-5)
mynorm *= llm.model.layers[2].input_layernorm.weight
mynorm = mynorm.to(torch.bfloat16)
print('llm: ',(norm - mynorm).abs().mean())
# input layer norm is only applied to hidden_states argument https://github.com/vllm-project/vllm/blob/bc6ccb987877000ec271e0076317b03a66cde4bc/vllm/model_executor/models/llama.py#L274
# vllm carries around hidden_states and residual separately
# following the steps from here https://github.com/vllm-project/vllm/blob/bc6ccb987877000ec271e0076317b03a66cde4bc/vllm/model_executor/layers/layernorm.py#L38
myvtmp = vresidual[0]
myvnorm = myvtmp.float() * torch.rsqrt(myvtmp.float().pow(2).mean(dim=-1, keepdim=True) + 1e-5)
myvnorm *= vllm.model.layers[2].input_layernorm.weight
myvnorm = myvnorm.to(torch.bfloat16)
print('vllm: ',(vnorm - myvnorm).abs().mean())
#myvnorm
```

**Comment by MichaelRipa** [comment]
Just confirming that I was able to replicate this issue locally. Will look into it!

**Comment by MichaelRipa** [comment]
Let me make the case for why I don’t think this is an NNsight-related bug.

Given the various optimizations implemented by VLLM, it seems likely that the internal representations and latents are modified in a non-trivial way. The fact that the layer norm matches doesn’t surprise me, as it aggregates values and normalizes them back to the mean, likely a key factor in preserving output integrity.

If this were an NNsight bug, it would most likely stem from an indexing issue since NNsight hooks into the model. Conversely, if the differences were purely intrinsic, we’d expect discrepancies across all layers.

What I find is that while all the inputs and outputs between `VLLM` and `LanguageModel` differ, the deviations grow progressively larger as we move deeper into the model.

Here’s the mean difference between the inputs (outputs follow a similar pattern):

```
Layer 0: 0.0048828125
Layer 1: 0.015380859375
Layer 2: 0.1123046875
Layer 3: 0.11962890625
Layer 4: 0.130859375
Layer 5: 0.138671875
Layer 6: 0.1474609375
Layer 7: 0.1591796875
Layer 8: 0.171875
Layer 9: 0.1796875
Layer 10: 0.1826171875
Layer 11: 0.1865234375
Layer 12: 0.1943359375
Layer 13: 0.1962890625
Layer 14: 0.19921875
Layer 15: 0.20703125
Layer 16: 0.21875
Layer 17: 0.232421875
Layer 18: 0.2451171875
Layer 19: 0.2578125
Layer 20: 0.267578125
Layer 21: 0.283203125
Layer 22: 0.3046875
Layer 23: 0.3203125
Layer 24: 0.333984375
Layer 25: 0.34765625
Layer 26: 0.3671875
Layer 27: 0.3828125
Layer 28: 0.40625
Layer 29: 0.4375
Layer 30: 0.48828125
Layer 31: 0.578125
```

However, if we swap indices between the two models using `.input[0]` for `LanguageModel` and `.input[1]` for `VLLM`, the differences no longer significantly increase monotonically:

```
Layer 0: 1.0
Layer 1: 1.0
Layer 2: 1.0703125
Layer 3: 1.0703125
Layer 4: 1.0703125
Layer 5: 1.0703125
Layer 6: 1.0703125
Layer 7: 1.0703125
Layer 8: 1.0703125
Layer 9: 1.0703125
Layer 10: 1.0703125
Layer 11: 1.0703125
Layer 12: 1.0703125
Layer 13: 1.0703125
Layer 14: 1.0703125
Layer 15: 1.0703125
Layer 16: 1.0703125
Layer 17: 1.0703125
Layer 18: 1.0703125
Layer 19: 1.0703125
Layer 20: 1.0703125
Layer 21: 1.0703125
Layer 22: 1.0703125
Layer 23: 1.0703125
Layer 24: 1.0703125
Layer 25: 1.078125
Layer 26: 1.078125
Layer 27: 1.0859375
Layer 28: 1.0859375
Layer 29: 1.0859375
Layer 30: 1.09375
Layer 31: 1.109375
```

This suggests that the indexing is correct.

(I know we discussed some of this offline, feel free to share any additional thoughts here.)
